src/main/java/com/faforever/client/filter/FilterBuilder.java

```diff
@@ -0,0 +1,85 @@
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.Range;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+
+@RequiredArgsConstructor(access = AccessLevel.MODULE)
+public class FilterBuilder<T> {
+
+  private final UiService uiService;
+  private final Consumer<AbstractFilterNodeController<?, ? extends ObservableValue<?>, T>> onFilterBuilt;
+
+  public FilterCheckboxController<T> checkbox(String text, BiFunction<Boolean, T, Boolean> filter) {
+    FilterCheckboxController<T> controller = uiService.loadFxml("theme/filter/checkbox_filter.fxml");
+    controller.setText(text);
+    controller.registerListener(filter);
+    onFilterBuilt.accept(controller);
+    return controller;
+  }
+
+  public FilterTextFieldController<T> textField(String promptText, BiFunction<String, T, Boolean> filter) {
+    FilterTextFieldController<T> controller = uiService.loadFxml("theme/filter/textfield_filter.fxml");
+    controller.setPromptText(promptText);
+    controller.registerListener(filter);
+    onFilterBuilt.accept(controller);
+    return controller;
+  }
+
+  public <U> FilterMultiCheckboxController<U, T> multiCheckbox(String text, List<U> items, StringConverter<U> converter, BiFunction<List<U>, T, Boolean> filter) {
+    return multiCheckbox(text, CompletableFuture.completedFuture(items), converter, filter);
+  }
+
+  public <U> FilterMultiCheckboxController<U, T> multiCheckbox(String text, CompletableFuture<List<U>> future, StringConverter<U> converter, BiFunction<List<U>, T, Boolean> filter) {
+    FilterMultiCheckboxController<U, T> controller = uiService.loadFxml("theme/filter/multicheckbox_filter.fxml");
+    controller.setText(text);
+    controller.setConverter(converter);
+    future.thenAccept(items -> {
+      controller.setItems(items);
+      controller.registerListener(filter);
+    });
+    onFilterBuilt.accept(controller);
+    return controller;
+  }
+
+  public RangeSliderFilterController<T> rangeSlider(String text, double minValue, double maxValue, BiFunction<Range<Integer>, T, Boolean> filter) {
+    RangeSliderFilterController<T> controller = uiService.loadFxml("theme/filter/range_slider_filter.fxml", RangeSliderFilterController.class);
+    controller.setText(text);
+    controller.setMinValue(minValue);
+    controller.setMaxValue(maxValue);
+    controller.registerListener(filter);
+    onFilterBuilt.accept(controller);
+    return controller;
+  }
+
+  public <I> RangeSliderWithChoiceFilterController<I, T> rangeSliderWithCombobox(String text, CompletableFuture<List<I>> future, StringConverter<I> converter, double minValue, double maxValue, BiFunction<ItemWithRange<I, Integer>, T, Boolean> filter) {
+    RangeSliderWithChoiceFilterController<I, T> controller = uiService.loadFxml("theme/filter/range_slider_filter.fxml", RangeSliderWithChoiceFilterController.class);
```

Comment:
I don't particularly like that this takes a future as input but I see why you did and can't think necessarily of a better way besides blocking the construction of the filter which might be better

Reply:
Uhhh, I do not want to redo the class for the nth time. :D